### ISSUE TITLE
Parsing error raised by supabase stripe datawrapper

### ISSUE THREAD

**Issue by Goldziher** [sqlfluff/sqlfluff#6027]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

```text
➜  supabase git:(main) ✗ pre-commit run --all-files
sqlfluff-fix.............................................................Failed
- hook id: sqlfluff-fix
- exit code: 1

==== finding fixable violations ====
  [1 templating/parsing errors found]
```

Tracing this error, it seems that the following file (based on the supabase documentation) is causing the issue:

```sql
-- =============================================
-- Stripe Foreign Data Wrapper setup
-- =============================================

-- Create the wrappers extension if it doesn't exist
create extension if not exists wrappers with schema extensions;

-- Create the Stripe foreign data wrapper
create foreign data wrapper stripe_wrapper
    handler stripe_fdw_handler
    validator stripe_fdw_validator;

-- Save Stripe API key in Vault and retrieve the key_id
-- Replace <API_KEY> with your actual Stripe API key
insert into vault.secrets (name, secret)
values (
           'stripe',
           '<API_KEY>'
       )
returning key_id;

-- Create the Stripe server
-- Replace <key_ID> with the actual Key ID from the above insert
create server stripe_server
    foreign data wrapper stripe_wrapper
    options (
        api_key_id '<key_ID>',
        api_url 'https://api.stripe.com/v1/',  -- Stripe API base URL
        api_version '2024-06-20'  -- Stripe API version
        );

-- =============================================
-- Stripe schema and table definitions
-- =============================================

-- Create Stripe schema
create schema stripe;

-- Stripe accounts table
-- Stores information about Stripe accounts
create foreign table stripe.accounts (
    id text,
    business_type text,
    country text,
    email text,
    type text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'accounts'
        );

-- Stripe balance table
-- Stores balance information
create foreign table stripe.balance (
    balance_type text,
    amount bigint,
    currency text,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'balance'
        );

-- Stripe balance transactions table
-- Stores information about balance transactions
create foreign table stripe.balance_transactions (
    id text,
    amount bigint,
    currency text,
    description text,
    fee bigint,
    net bigint,
    status text,
    type text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'balance_transactions'
        );

-- Stripe charges table
-- Stores information about charges
create foreign table stripe.charges (
    id text,
    amount bigint,
    currency text,
    customer text,
    description text,
    invoice text,
    payment_intent text,
    status text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'charges'
        );

-- Stripe checkout sessions table
-- Stores information about checkout sessions
create foreign table stripe.checkout_sessions (
    id text,
    customer text,
    payment_intent text,
    subscription text,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'checkout/sessions',
        rowid_column 'id'
        );

-- Stripe customers table
-- Stores information about customers
create foreign table stripe.customers (
    id text,
    email text,
    name text,
    description text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'customers',
        rowid_column 'id'
        );

-- Stripe disputes table
-- Stores information about disputes
create foreign table stripe.disputes (
    id text,
    amount bigint,
    currency text,
    charge text,
    payment_intent text,
    reason text,
    status text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'disputes'
        );

-- Stripe events table
-- Stores information about Stripe events
create foreign table stripe.events (
    id text,
    type text,
    api_version text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'events'
        );

-- Stripe files table
-- Stores information about files
create foreign table stripe.files (
    id text,
    filename text,
    purpose text,
    title text,
    size bigint,
    type text,
    url text,
    created timestamp,
    expires_at timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'files'
        );

-- Stripe file links table
-- Stores information about file links
create foreign table stripe.file_links (
    id text,
    file text,
    url text,
    created timestamp,
    expired bool,
    expires_at timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'file_links'
        );

-- Stripe invoices table
-- Stores information about invoices
create foreign table stripe.invoices (
    id text,
    customer text,
    subscription text,
    status text,
    total bigint,
    currency text,
    period_start timestamp,
    period_end timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'invoices'
        );

-- Stripe mandates table
-- Stores information about mandates
create foreign table stripe.mandates (
    id text,
    payment_method text,
    status text,
    type text,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'mandates'
        );

-- Stripe payment intents table
-- Stores information about payment intents
create foreign table stripe.payment_intents (
    id text,
    customer text,
    amount bigint,
    currency text,
    payment_method text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'payment_intents'
        );

-- Stripe payouts table
-- Stores information about payouts
create foreign table stripe.payouts (
    id text,
    amount bigint,
    currency text,
    arrival_date timestamp,
    description text,
    statement_descriptor text,
    status text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'payouts'
        );

-- Stripe prices table
-- Stores information about prices
create foreign table stripe.prices (
    id text,
    active bool,
    currency text,
    product text,
    unit_amount bigint,
    type text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'prices'
        );

-- Stripe products table
-- Stores information about products
create foreign table stripe.products (
    id text,
    name text,
    active bool,
    default_price text,
    description text,
    created timestamp,
    updated timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'products',
        rowid_column 'id'
        );

-- Stripe refunds table
-- Stores information about refunds
create foreign table stripe.refunds (
    id text,
    amount bigint,
    currency text,
    charge text,
    payment_intent text,
    reason text,
    status text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'refunds'
        );

-- Stripe setup attempts table
-- Stores information about setup attempts
create foreign table stripe.setup_attempts (
    id text,
    application text,
    customer text,
    on_behalf_of text,
    payment_method text,
    setup_intent text,
    status text,
    usage text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'setup_attempts'
        );

-- Stripe setup intents table
-- Stores information about setup intents
create foreign table stripe.setup_intents (
    id text,
    client_secret text,
    customer text,
    description text,
    payment_method text,
    status text,
    usage text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'setup_intents'
        );

-- Stripe subscriptions table
-- Stores information about subscriptions
create foreign table stripe.subscriptions (
    id text,
    customer text,
    currency text,
    current_period_start timestamp,
    current_period_end timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'subscriptions',
        rowid_column 'id'
        );

-- Stripe tokens table
-- Stores information about tokens
create foreign table stripe.tokens (
    id text,
    customer text,
    currency text,
    current_period_start timestamp,
    current_period_end timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'tokens'
        );

-- Stripe topups table
-- Stores information about top-ups
create foreign table stripe.topups (
    id text,
    amount bigint,
    currency text,
    description text,
    status text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'topups'
        );

-- Stripe transfers table
-- Stores information about transfers
create foreign table stripe.transfers (
    id text,
    amount bigint,
    currency text,
    description text,
    destination text,
    created timestamp,
    attrs jsonb
    )
    server stripe_server
    options (
        object 'transfers'
        );
```

The part that is causing the tool to fail is:

```sql
create foreign data wrapper stripe_wrapper
    handler stripe_fdw_handler
    validator stripe_fdw_validator;
```

Removing this allows the tool to run and validate the file.

Importantly- after failing to process the file, you still get a pass message in the pre-commit CLI:

```text
sqlfluff-fix.............................................................Failed
- hook id: sqlfluff-fix
- exit code: 1

==== finding fixable violations ====
  [1 templating/parsing errors found]
==== no fixable linting violations found ====
All Finished 📜 🎉!
==== lint for unfixable violations ====
== [supabase/schema.sql] PASS
== [supabase/seed.sql] PASS
== [supabase/stripe.sql] PASS
```

### Expected Behaviour

The tool is able to gracefully handle unknown syntax and doesn't fail to process the entire file. Perhaps an error is raised that can be suppressed. 

### Observed Behaviour

The tool fails with a templating error, but gives no input on why. It also doesn't process the file after the error occurred, but still gives a gree pass in pre-commit (without fixing any issues in the file).

### How to reproduce

execute the sqlfluff tool against the give schema 

### Dialect

postgres

### Version

3.1.0 (pre-commit)

### Configuration

```toml
[tool.sqlfluff.core]
dialect = "postgres"
max_line_length = 120
exclude_rules = ["RF05"]

[tool.sqlfluff.indent_unit]
indent_unit = "space"
tab_space_size = 4

[tool.sqlfluff.rules.capitalisation.keywords]
capitalisation_policy = "lower"

[tool.sqlfluff.rules.capitalisation.identifiers]
capitalisation_policy = "lower"

[tool.sqlfluff.rules.capitalisation.functions]
capitalisation_policy = "lower"

[tool.sqlfluff.rules.capitalisation.literals]
capitalisation_policy = "lower"

[tool.sqlfluff.rules.capitalisation.types]
capitalisation_policy = "lower"

[tool.sqlfluff.rules.layout.long_lines]
max_line_length = 120

[tool.sqlfluff.rules.structure.column_order]
order_by_style = "explicit"
```

Im willing to submit a PR, but I am no expert in this tool at all. I am very competent with Python though.

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
You've raised a couple of issues here.

Firstly, on the intent of the pre-commit hooks. The `fix` hook is designed to pass when it's successfully fixed what it can - not to fail when it finds an issue. That means when it finds a parsing error, it doesn't return a fail, it just skips the file because no fixes are possible. In your case you might want to use the `lint` hook instead - which will also fail on parsing errors.

Secondly on processing the rest of the file - in some cases we are able to do this, but in many cases we can't because a parsing error earlier in the file may prevent understanding the rest of the file in context. 

You've also highlighted a missing syntax for the `postgres` dialect. If you're up for helping to resolve that, there's some fairly detailed documentation [on the SQLFluff wiki](https://github.com/sqlfluff/sqlfluff/wiki/Contributing-Dialect-Changes) about how to contribute dialect gaps. This is a really helpful way to support the project as you've got access to a database of the right flavour to check that what gets implemented actually works.

**Comment by Goldziher** [comment]
Sounds great. I want to find a time for helping with the syntax. Let me see if I can, otherwise - ill raise the issue with the supabase peeps, they might contribute as well.

**Comment by Goldziher** [comment]
> Secondly on processing the rest of the file - in some cases we are able to do this, but in many cases we can't because a parsing error earlier in the file may prevent understanding the rest of the file in context.

In this case, I would suggest the tool fails gracefully with a clear error message or warning. It took a bit of effort to understand the cause of the failure, since there was no output to console, it simply skipped over the file.
